### ISSUE TITLE
Cyclic dependency with Puppet 5 and recent versions of module

### ISSUE THREAD

**Issue by rcousens** [voxpupuli/puppet-nodejs#329]
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: v5.3.2
- Ruby: 2.4.1
- Distribution: Ubuntu 14.04 LTS
- Module version: 3.1.0/4.0.0

## How to reproduce (e.g Puppet code you use)

```
  class { 'nodejs':
    repo_url_suffix => '4.x',
  }
```

## What are you seeing

```
Error: Found 1 dependency cycle:
(Anchor[apt_key 9FD3B784BC1C6FC31A8A0A1C1655A0AB68576280 present] => Apt::Key[Add key: 9FD3B784BC1C6FC31A8A0A1C1655A0AB68576280 from Apt::Source nodesource] => Apt::Source[nodesource] => Package[apt-transport-https] => Apt::Source[nodesource] => Apt::Key[Add key: 9FD3B784BC1C6FC31A8A0A1C1655A0AB68576280 from Apt::Source nodesource] => Anchor[apt_key 9FD3B784BC1C6FC31A8A0A1C1655A0AB68576280 present])
```

## What behaviour did you expect instead

NodeJS installed

## Any additional information you'd like to impart

This is absolutely related to the require Package statements in the apt::source resource from repo/nodesource/apt.pp. I understand the necessity for the require as apt-transport-https must be present for apt-get update to work with the nodesource repo, but I am unsure how this creates a cyclic dependency.

**Comment by rcousens** [comment]
```
node default {

  Apt::Pin <| |> -> Package <| |>
  Apt::Source <| |> -> Package <| |>

  difference(lookup('classes', Array[String], 'unique'), lookup('rejected_classes', Array[String],'unique',[])).include
}
```

Removing the Apt::Source collector for packages fixes the problem. I feel like a dolt, however I also actually want Apt::Source to be applied before all packages are installed. Egg/chicken!

**Comment by rcousens** [comment]
I think this can be closed. I apologise for the fuss. I saw it was a previous problem with Puppet 3 and thought it was rearing its ugly head again. For me, the issue came solely because I had a site manifest using an Apt::Source resource collector to express a dependency on processing packages afterwards. 

I've heard resource collectors are hacky and unwieldy and will bite you in the ass, I guess I just found out why! They override a lot of internal logic expressed in modules themselves. Thanks!
